fplus-lib/src/external_services/similarity_detection.rs

```diff
@@ -0,0 +1,182 @@
+            client_address: app.client_address.clone(),
+            text: app.application.data_set_sample.clone(),
+        });
+        if comparable_data.data_owner_name == app.application.data_owner_name {
+            existing_data_owner_name.push(app.client_address.clone());
+        }
+    }
+    let similar_project_desciptions = get_similar_texts_tfidf(&projects_descriptions)?;
+    let similar_stored_data_desciptions = get_similar_texts_tfidf(&stored_data_descriptions)?;
+    let similar_project_and_stored_data_desciptions =
+        get_similar_texts_tfidf(&projects_and_stored_data_descriptions)?;
+    let similar_data_set_sample = get_similar_texts_levenshtein(&data_set_samples)?;
+
+    let unique_addresses: HashSet<String> = similar_project_desciptions
+        .unwrap_or_default()
+        .into_iter()
+        .chain(
+            similar_stored_data_desciptions
+                .unwrap_or_default()
+                .into_iter(),
+        )
+        .chain(
+            similar_project_and_stored_data_desciptions
+                .unwrap_or_default()
+                .into_iter(),
+        )
+        .chain(similar_data_set_sample.unwrap_or_default().into_iter())
+        .collect();
+    let unique_addresses: Vec<String> = unique_addresses.into_iter().collect();
+
+    let _applications = get_applications_by_clients_addresses(unique_addresses)
+        .await
+        .map_err(|e| LDNError::New(format!("Failed to get applications from database: {}", e)))?;
+    Ok(())
+}
+
+fn get_similar_texts_tfidf(documents: &Vec<Document>) -> Result<Option<Vec<String>>, LDNError> {
+    let mut tokenized_documents = Vec::new();
+    for doc in documents {
+        tokenized_documents.push(tfidf_summarizer::tokenize(&doc.text));
+    }
+    let df = tfidf_summarizer::document_frequency(&tokenized_documents);
+    let documents_words: Vec<String> = df.keys().cloned().collect();
+    let idf = tfidf_summarizer::inverse_document_frequency(&df, tokenized_documents.len());
+    let mut tfidf_result = Vec::new();
+
+    for tokens in tokenized_documents.iter() {
+        tfidf_result.push(tfidf_summarizer::tf_idf(tokens.clone(), &idf));
+    }
+    let documents_converted_to_array = convert_to_ndarray(&tfidf_result, &documents_words);
+    let mut similar_applications: Vec<String> = Vec::new();
+    let tfidf_threshold = get_env_var_or_default("TFIDF_THRESHOLD")
+        .parse::<f64>()
+        .map_err(|e| LDNError::New(format!("Parse tfidf threshold score to f64 failed: {}", e)))?;
+    for i in 1..documents_converted_to_array.len() {
+        let similarity = cosine_similarity(
+            &documents_converted_to_array[0],
+            &documents_converted_to_array[i],
+        );
+        if similarity > tfidf_threshold {
+            similar_applications.push(documents[i].client_address.clone());
+        }
+    }
+    if similar_applications.is_empty() {
+        return Ok(None);
+    }
+    Ok(Some(similar_applications))
```

Comment:
```suggestion
    Ok(similar_applications)
```
Just return the empty vec, no point in wrapping it in `Option` in this case.